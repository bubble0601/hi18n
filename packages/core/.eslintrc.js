module.exports = /** @type {import("eslint").Linter.Config} */ ({
  parserOptions: {
    project: [
      require.resolve("./tsconfig.json"),
      require.resolve("./configs/tsconfig.main.json"),
    ],
  },
});
