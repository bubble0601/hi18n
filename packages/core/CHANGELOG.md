## 0.1.1

- Add `translationId` and `t.dynamic` for dynamically selecting translations
- Add `t.todo` for bootstrapping new translations

## 0.1.0

Initial release.
