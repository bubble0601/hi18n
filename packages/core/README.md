## Core library for [hi18n](https://github.com/wantedly/hi18n)

See [hi18n's README](https://github.com/wantedly/hi18n#readme) for general information.

## msg

Returns the first argument. Used for type-safe translations.

```typescript
export default new Catalog<Vocabulary>("en", {
  "example/greeting": msg("Hello, {name}!"),
});
```

## Catalog

Creates a set of translated messages for a specific locale.

```typescript
export default new Catalog<Vocabulary>("en", {
  "example/greeting": msg("Hello, {name}!"),
});
```

## Book

Creates a set of translated messages for all supported locales.

```typescript
const book = new Book<Vocabulary>({
  en: catalogEn,
  ja: catalogJa,
});
```

## getTranslator

Retrieves a translation helper from a book and a locale.

```typescript
const { t } = getTranslator(book, "ja");
t("example/greeting", { name: "太郎" });
```
