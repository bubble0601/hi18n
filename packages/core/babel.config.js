module.exports = {
  targets: "defaults, not ie 11, not ie_mob 11",
  presets: [
    ["@babel/preset-env", { modules: "commonjs" }],
    ["@babel/preset-typescript", { allowDeclareFields: true }],
  ],
  plugins: [
    ["@babel/plugin-transform-runtime", { version: "^7.18.3" }],
    [
      "babel-plugin-polyfill-corejs3",
      { method: "usage-pure", version: "^3.23.2" },
    ],
  ],
};
