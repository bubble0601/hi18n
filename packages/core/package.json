{
  "name": "@hi18n/core",
  "version": "0.1.0",
  "description": "Stateless message catalog for internationalization",
  "main": "dist/index.js",
  "repository": "https://github.com/qnighy/hi18n",
  "author": "Masaki Hara <ackie.h.gmai@gmail.com>",
  "license": "MIT",
  "packageManager": "yarn@3.1.0",
  "scripts": {
    "build": "tsc --build",
    "test": "jest"
  },
  "dependencies": {
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "tslib": "^2.3.1"
  },
  "devDependencies": {
    "@babel/core": "^7.16.5",
    "@babel/preset-env": "^7.16.5",
    "@babel/preset-react": "^7.16.5",
    "@babel/preset-typescript": "^7.16.5",
    "@jest/globals": "^27.4.4",
    "@testing-library/jest-dom": "^5.16.3",
    "@testing-library/react": "^12.1.4",
    "@types/react": "^17.0.43",
    "@types/react-dom": "^17.0.11",
    "@types/testing-library__jest-dom": "^5.14.2",
    "babel-jest": "^27.4.5",
    "expect": "^27.4.2",
    "jest": "^27.4.5",
    "ts-jest": "^27.1.4",
    "typescript": "^4.5.4"
  },
  "jest": {
    "testEnvironment": "jsdom"
  }
}
