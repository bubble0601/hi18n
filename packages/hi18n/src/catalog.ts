import { File, Program } from "@babel/types";
import traverse from "@babel/traverse";
import { enrichContext, EnrichOptions } from "./enrich-traverse";

export function getCatalog(code: string, program: File | Program, options: EnrichOptions = {}) {
  const { scope } = enrichContext(code, program, options);
  traverse(program, {
    ImportDeclaration(path) {
      if (path.node.source.value !== "@hi18n/core") return;
      // for (const specifier of path.get("specifiers")) {
      //
      // }
    },
    ExportDeclaration(path) {
      if (path.node.type === "ExportDefaultDeclaration") return;
      if (path.node.source?.value !== "@hi18n/core") return;
      throw path.buildCodeFrameError("Cannot track this usage");
    },
  }, scope);
}
