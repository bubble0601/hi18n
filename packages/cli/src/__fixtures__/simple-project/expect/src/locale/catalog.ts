import { Message, MessageCatalog } from "@hi18n/core";
import catalogEn from "./catalog-en";
import catalogJa from "./catalog-ja";

export type Messages = {
  "example/greeting": Message,
  "example/greeting-todo-newtranslation": Message;
  // "example/greeting-unused": Message,
  "example/greeting-todo": Message,
};

export const catalog = new MessageCatalog<Messages>({
  en: catalogEn,
  ja: catalogJa,
});
