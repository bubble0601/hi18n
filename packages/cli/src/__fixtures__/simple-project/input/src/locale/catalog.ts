import { Message, Book } from "@hi18n/core";
import catalogEn from "./catalog-en";
import catalogJa from "./catalog-ja";

export type Vocabulary = {
  "example/greeting": Message,
  "example/greeting-unused": Message,
  // "example/greeting-todo": Message,
};

export const book = new Book<Vocabulary>({
  en: catalogEn,
  ja: catalogJa,
});
