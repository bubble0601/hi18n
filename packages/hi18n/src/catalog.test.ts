import { describe, expect, it } from "@jest/globals";
import { getCatalog } from "./catalog";
import { parseAsync } from "@babel/core";

describe("getCatalog", () => {
  describe("export declarations", () => {
    it("throws error on untrackable re-exports", async () => {
      const code = `
export * from "@hi18n/core";
`;
      const program = await parseAsync(code, { sourceType: "module", configFile: false });
      expect(() => {
        getCatalog(code, program!, { highlightCode: false });
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
