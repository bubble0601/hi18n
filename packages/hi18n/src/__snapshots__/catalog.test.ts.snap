// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCatalog export declarations throws error on untrackable re-exports 1`] = `
"Cannot track this usage
  1 |
> 2 | export * from \\"@hi18n/core\\";
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  3 |"
`;
